### DISCUSSION TITLE
Adding a bitmap to the file coms

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1856] [Q&A]
For my WebCard project I need to add a bitmap to the coms of the program. I have a general idea how to start but I'm not sure how to insert the actual data of the bitmap in the coms.

Assume the variable `MyBitMap` holds the bitmap. One way to proceed is to evaluate `(ADDTOFILE '(MyBitMap <value>) 'VARS 'MYCOMS)` at an Interlisp Exec. But how do I insert the bitmap data at `<value>`?

I tried editing the coms with `(DC MYPRO)` and doing something like `(ADDTOFILE '(MyBitMap ,MYBitMap) 'VARS 'MYPROG)` but SEdit renders it as `(ADDTOFILE '(MyBitMap %,MYBitMap) 'VARS 'MYPROG)` with no data. `(BITMAPS MyBitMap)` likely goes in the coms too.

Any advice?

**Comment by nbriggs** [comment]
If `MyBitmap` has a value, then adding `(BITMAPS MyBitmap)` to the coms will save the current value when you do the `MAKEFILE` (or `CLEANUP`).  In fact, you don't even need to tell it that it's a bitmap, `(VARS MyBitmap)` will work just as well.

**Reply by pamoroso** [reply]
Okay thanks, I'll go with `(BITMAPS MyBitMap)`. I assume to access the bitmap from the program I can just reference the variable `MyBitMap`.

**Reply by nbriggs** [reply]
Yes, it's just a regular variable with a value.

**Reply by nbriggs** [reply]
You could also have done:

`(ADDTOFILE (LIST 'MyBitmap MyBitmap) 'BITMAPS 'MYPRO))`

Assuming the file is called `MYPRO`.  It'll add it to `MYPROCOMS` and when the file gets printed by `MAKEFILE` it'll print the bitmap value.  Think carefully about what'll happen if you change the value of MyBitmap to some other bitmap - the MYPROCOMS has a reference to the `BITMAP` object that was the value of MyBitmap at the time you executed the command, if you rebound MyBitmap you wouldn't get the new value, but if you destructively edited the old value that *will* affect the value saved when you execute the MAKEFILE.

Another way to do it, in SEDIT, is to `(DC MYPRO)`, and edit in by hand, for example --

`((BITMAPS (MyBitmap MyBitmap)))`

then, assuming `MyBitmap` has the value you want, select the second one and M-Z (mutate) by EVAL.  It will be replaced by its value, and then when you MAKEFILE the same thing will happen as above, with the same caveats.

**Reply by pamoroso** [reply]
For this project I have a bitmap ready I don't anticipate changing but in case I'll keep an eye out, thanks.

**Reply by pamoroso** [reply]
I tried to add the bitmap but without success.

First, I assigned the bitmap data to `WCD.WebCardBitMap`. Next, I evaluated `(ADDTOFILE (LIST 'WCD.WebCardBitMap WCD.WebCardBitMap) 'BITMAPS 'WEBCARD)`, and referenced `WCD.WebCardBitMap` from function `WCD.CreateWebCardType` to use the bitmap:

```
(WCD.CreateWebCardType
  [LAMBDA NIL                                                (* ; "Edited 21-Oct-2024 13:01 by PA")
                                                             (* ; "Edited 14-Oct-2024 11:55 by PA")
                                                             (* ; "Edited  1-Oct-2024 17:35 by PA")
                                                             (* Define the new card type Web that 
                                                             inherits from Text.)
    (DECLARE (GLOBALVARS WCD.WebCardBitMap))
    (NCP.CreateCardType 'Web 'Text '((MakeFn WCD.MakeWebCard))
           `((DefaultHeight 70)
             (LinkAnchorModesSupported NIL)
             (DisplayedInMenuFlg T)
             (LinkIconAttachedBitMap ,WCD.WebCardBitMap])
```

Finally I saved with `(MAKEFILE 'WEBCARD`). But now when I load `WEBCARD` I get the error `WCD.WebCardBitMap is an unbound variable.`

Do I need to also add `WCD.WebCardBitMap` as a global variable with something like `(ADDTOFILE (LIST 'WCD.WebCardBitMap WCD.WebCardBitMap) 'VARS 'WEBCARD)`?

Here is the full code of WEBCARD so far: [WEBCARD.txt](https://github.com/user-attachments/files/17459830/WEBCARD.txt)

**Comment by nbriggs** [comment]
Look at the COMS (or look at the file) -- you're evaluating the `(WCD.InitWebCard)` before you define the value of WCD.WebCardBitMap (your "P" command precedes the "BITMAPS" command).

The order of commands in the filecoms control the order of things written (and thus read, and executed), so the order is important -- always a good idea to look at them with, say, `(DC WEBCARD)` to make sure they make sense.

**Comment by pamoroso** [comment]
I moved the `P` command to the end of the coms but `(FILES?)` now asks where to put all the functions I already assigned to the `WEBCARD` file:

![webcardcoms](https://github.com/user-attachments/assets/772f419a-36e6-46ec-a731-cdf284a0b733)

Is this expected when rearranging the coms?

**Reply by nbriggs** [reply]
That's not what I would have expected.  Say `N` and see what happens.

**Reply by pamoroso** [reply]
Answering `N` to `FILES?` to skip the function assignment seems to yield the expected coms with all the functions, so it's probably safe to `MAKEFILE`:

![WEBCARDCOMS2](https://github.com/user-attachments/assets/f627cd93-07ac-4907-8132-3cfed4f753ff)

**Reply by nbriggs** [reply]
You’re doing it on a versioned file system, so previous versions are there. Don’t keep exiting your Medley session though.

**Comment by nbriggs** [comment]
When I did the same, starting from your WEBCARD(.txt), and moved the P command it just said that WEBCARDCOMS had changed, nothing else.

**Comment by pamoroso** [comment]
`(MAKEFILE 'WEBCARD)` saved the file with `P` at the end of the coms but printed the warning `(no LinkIconLeftButtonFn property for Web)`. Here's the latest file:  [WEBCARD.txt](https://github.com/user-attachments/files/17474718/WEBCARD.txt)

Loading the modified file completes with no errors and the bitmap can now be seen as expected in link icons leading to cards of type Web.

However, traversing such a link no longer opens the web browser to visit the link for the reason hinted by the `MAKEFILE` warning. `MAKEFILE` skipped the `LinkIconLeftButtonFn` property of the symbol `Web` which [was present](https://github.com/pamoroso/webcard/blob/791a9924b2ccd4c68ea7a0acd319f2a21c90697c/WEBCARD#L200) in the last working version of the program prior to my attempts to add a bitmap. [The function](https://github.com/pamoroso/webcard/blob/791a9924b2ccd4c68ea7a0acd319f2a21c90697c/WEBCARD#L111) is called when left-clicking a Web link icon.

**Comment by nbriggs** [comment]
At the time you did the MAKEFILE there was no LinkIconLeftButtonFn property on the atom Web, presumably as a result of the LOAD erroring before the property got restored.
```
        (P (WCD.InitWebCard))
        (PROP LinkIconLeftButtonFn Web)
        (BITMAPS (WCD.WebCardBitMap ...))
```

So there was nothing to save, and it warned you about it.  The PROP file package command is designed more for saving a property value where you don't know what the value should be. I'd suggest explicitly putting the known property value on `Web` as part of creating the Web card type.

**Reply by pamoroso** [reply]
I'll do this. I assume it's better to also remove `PROP` from the coms, right?

**Reply by nbriggs** [reply]
Yes, remove the PROP from the coms.

**Reply by pamoroso** [reply]
Along with the prior bitmap change in the coms I removed `PROP` from the coms and [moved the property assignment to `WCD.CreateWebcardType`](https://github.com/pamoroso/webcard/blob/b926f67169cf23cde169acf9ad3c5f078e414e65/WEBCARD#L67). Now everything loads and works fine. Thanks.

**Comment by MattHeffron** [comment]
@pamoroso Is this OK to close now?

**Comment by pamoroso** [comment]
@MattHeffron Yes, I'm closing the discussion.
